OSDOCS-4812: Creating procedure for extending edge nodes into AWS LZ (#57063)
openshift/openshift-docs: modules/installation-extend-edge-nodes-aws-local-zones.adoc

ifeval::["{context}" == "installing-aws-localzone"]
:localzone:
endif::[]

:_content-type: PROCEDURE
[id="installation-extend-edge-nodes-aws-local-zones_{context}"]
= Creating user workloads in AWS Local Zones
After you create an Amazon Web Service (AWS) Local Zone environment, and you deploy your cluster, you can use edge worker nodes to create a user workloads in Local Zone subnets. 

After the `openshift-installer` creates the cluster, the installer automatically specifies a taint effect of `NoSchedule` to each edge worker node. This means that a scheduler does not add a new pod to a node if the pod's deployment does not match the specified tolerations for a taint. You can modify the taint for better control over how each node creates a workload in each Local Zone subnet.

The `openshift-installer` creates the Machine Set manifest and attaches `node-role.kubernetes.io/edge` and `node-role.kubernetes.io/worker` labels to each edge worker node that is located in a Local Zone subnet.

.Prerequisites

* You have an access to the OpenShift CLI (`oc`).
* You configured an AWS Local Zones environment. 
* You deployed your cluster om a compatible cloud platform. 

> * You ensured that your worker manifest file specifies `node-role.kubernetes.io/edge` for the `taint` parameter.

Author: Well, spotted. Updated.

+
.Example of a configured `deployment` resource for an edge worker node that operates in a Local Zone subnet
[source,yaml]
----
apiVersion: apps/v1
kind: Deployment <1>
metadata:
  name: <local_zone_application> <2>
  namespace: <local_zone_application_namespace> <3>
spec:
  selector: 
    matchLabels:
      app: <local_zone_application>
  replicas: 1
  template:
    metadata:
      labels: 
        app: <local_zone_application>
        machine.openshift.io/zone-group: ${ZONE_GROUP_NAME} <4>
    spec:
      nodeSelector: <5>
        machine.openshift.io/zone-group: ${ZONE_GROUP_NAME}
      tolerations: <6>
      - key: "node-role.kubernetes.io/edge"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
      containers:
        - image: openshift/origin-node
          command:
           - "/bin/socat"
          args:
            - TCP4-LISTEN:8080,reuseaddr,fork
            - EXEC:'/bin/bash -c \"printf \\\"HTTP/1.0 200 OK\r\n\r\n\\\"; sed -e \\\"/^\r/q\\\"\"'
          imagePullPolicy: Always
          name: echoserver

>           ports:
>             - containerPort: 8080

Reviewer: Missing the volume mount
```suggestion
          ports:
            - containerPort: 8080
          volumeMounts:
            - mountPath: "/mnt/storage"
              name: data
```

Author: Thanks. Added.

>     volumes:

Author: Added from discussion on https://github.com/openshift/openshift-docs/pull/57736#discussion_r1152135801

    - name: data 

>       ephemeral:

Reviewer: This example was based on ephemeral volumes[1]. I don't have a strong opinion on keeping the ephemeral or regular as an example. WDYT  @yunjiang29 ?

[1] https://docs.openshift.com/container-platform/4.12/storage/generic-ephemeral-vols.html

Reviewer 2: Hi @dfitzmau @mtulio , agree that the ephemeral volume is not a good example as the volume will be cleaned up when the deployment deleted. It's not a good example for user general usage scenarios (persistent volume). A general example should be using workload with pvcs. We could replace the data volume `volume[0]` with persistent volume.
E.g.
```console
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: mypvc
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: gp2-csi
  volumeMode: Filesystem
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mydeployment
spec:
  selector:
    matchLabels:
      app: myapp
  replicas: 1
  template:
    metadata:
      labels:
        app: myapp
    spec:
      containers:
      - name: myc1
        image: quay.io/pewang/storagetest:base
        volumeMounts:
        - mountPath: /mnt/storage
          name: data
      volumes:
      - name: data
        persistentVolumeClaim:
          claimName: mypvc

```

Author: Thanks, @Phaow and @mtulio . I updated the example resource. I retained the `spec` stanza, because it details tolerations and so on. WDYT?

Reviewer 2: looks good to me. Thanks for the update!

Author: Hi @mtulio . Are you OK with this spec update? I can then send this PR onto the next review stage.

Reviewer: @dfitzmau for sure, I am ok (ty @Phaow ).

I would like to validate the entire flow in the p1 and p2 clusters, but it will not block you to move on to the next review stage.

        volumeClaimTemplate:
          metadata:
            labels:
              type: <application_ephemeral_volume>
          spec:
            accessModes: [ "ReadWriteOnce" ]

>             storageClassName: "gp2-csi"

Reviewer: @dfitzmau please add one legend item explaining this field, that is what we need to exemplify to our customers. Something like: 

> sotrageClassName: change to use `gp2-csi` when the Local Zone does not support `gp3-csi`

// Module included in the following assemblies:
//
// * post_installation_configuration/cluster-tasks.adoc

ifeval::["{context}" == "installing-aws-localzone"]
:localzone:
endif::[]

:_content-type: PROCEDURE
[id="installation-extend-edge-nodes-aws-local-zones_{context}"]
= Creating user workloads in AWS Local Zones
After you create an Amazon Web Service (AWS) Local Zone environment, and you deploy your cluster, you can use edge worker nodes to create a user workloads in Local Zone subnets. 

> After the `openshift-installer` creates the cluster, the installer automatically specifies a taint effect of `NoSchedule` to each edge worker node. This means that a scheduler does not add a new pod, or deployment, to a node if the pod does not match the specified tolerations for a taint. You can modify the taint for better control over how each node creates a workload in each Local Zone subnet.

Reviewer 3: ```suggestion
After the `openshift-installer` creates the cluster, the installation program automatically specifies a taint effect of `NoSchedule` to each edge worker node. This means that a scheduler does not add a new pod, or deployment, to a node if the pod does not match the specified tolerations for a taint. You can modify the taint for better control over how each node creates a workload in each Local Zone subnet.
```

The `openshift-installer` creates the machine set manifests file with `node-role.kubernetes.io/edge` and `node-role.kubernetes.io/worker` labels applied to each edge worker node that is located in a Local Zone subnet.

.Prerequisites

* You have an access to the OpenShift CLI (`oc`).
* You deployed your cluster in a Virtual Private Cloud (VPC) with defined Local Zone subnets.
* You ensured that the machine set for the edge workers on Local Zone subnets specifies the taints for `node-role.kubernetes.io/edge`.

.Procedure

. Create a `deployment` resource YAML file for an example application to be deployed in the edge worker node that operates in a Local Zone subnet. Ensure that you specify the correct tolerations that match the taints for the edge worker node.
+

> .Example of a configured `deployment` resource for an edge worker node that operates in a Local Zone subnet

Reviewer 4: @mtulio this demo will cause a warning, have you met this? can we ignore it?

```
persistentvolumeclaim/lz-pvc created
Warning: would violate PodSecurity "restricted:latest": seccompProfile (pod or container "echoserver" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
deployment.apps/lz-app created
```

Reviewer: Considering we must deliver the best practice on the docs, I think we need to fix it.

[source,yaml]
----
kind: PersistentVolumeClaim
apiVersion: v1
metadata:

>   name: <pvc_name>

Reviewer 4: @dfitzmau, it looks like the namespace is missing: `namespace: <local_zone_application_namespace>` for `PersistentVolumeClaim`.

In order to ensure consistency, we'd better to create namespace for this demo, like what @mtulio does in https://github.com/openshift/openshift-docs/pull/57063#pullrequestreview-1373220967

Reviewer: Good catch @yunjiang29  about the NS missing on PVC, I fixed it in my example and not commented here. I will provide a suggestion to fix it in a single-manifest file.

spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: gp2-csi <1>
  volumeMode: Filesystem

> ...

Reviewer 5: @dfitzmau, I've never seen a valid YAML that includes `...` before. Should this be `---` instead? This manifest contains two YAML objects? Thanks!

Author: Example snippet taken from https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

Author: @mtulio . Defining two YAML objects in one manifest file is OK? Should only one YAML object be in a manifest file?

Reviewer: It's fine, but we need to fix the `---`.
Please hold this PR, I am testing right now the entire example. I am getting some issues with indentation.

Reviewer 5: @dfitzmau, you can have multiple YAML definitions in a single file, but you need to separate them with `---`. I wasn't sure if your `...` was supposed to be a `---`.

Author: Thanks. Updated.

apiVersion: apps/v1
kind: Deployment <2>
metadata:
  name: <local_zone_application> <3>
  namespace: <local_zone_application_namespace> <4>
spec:
  selector: 
    matchLabels:
      app: <local_zone_application>
  replicas: 1
  template:
    metadata:
      labels: 
        app: <local_zone_application>

>         machine.openshift.io/zone-group: ${ZONE_GROUP_NAME} <5>

Author: Thanks, @mtulio . I should remove both references to `machine.openshift.io/zone-group`? 

![Screenshot from 2023-04-06 09-45-57](https://user-images.githubusercontent.com/57954076/230324524-360f08a0-da30-48e5-8988-99383cdfefdb.png)

    spec:
      nodeSelector: <6>
        machine.openshift.io/zone-group: ${ZONE_GROUP_NAME}
      tolerations: <7>
      - key: "node-role.kubernetes.io/edge"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
      containers:
        - image: openshift/origin-node
          command:
           - "/bin/socat"
          args:
            - TCP4-LISTEN:8080,reuseaddr,fork
            - EXEC:'/bin/bash -c \"printf \\\"HTTP/1.0 200 OK\r\n\r\n\\\"; sed -e \\\"/^\r/q\\\"\"'
          imagePullPolicy: Always
          name: echoserver
          ports:
            - containerPort: 8080
          volumeMounts:
            - mountPath: "/mnt/storage"
              name: data

>     volumes:
>     - name: data 
>       persistentVolumeClaim:
>         claimName: <pvc_name>

Reviewer: `volumes` should be under `podSpec`, otherwise, we'll get errors

```suggestion
      volumes:
      - name: data 
        persistentVolumeClaim:
          claimName: <pvc_name>
```

Author: Updated.

----
<1> `storageClassName`: For the Local Zone configuration, you must specify `gp2-csi`.
<2> `kind`: Defines the `deployment` resource.
<3> `name`: Specifies the name of your Local Zone application. For example, `local-zone-demo-app-nyc-1`.
<4> `namespace:` Defines the namespace for the AWS Local Zone where you want to run the user workload. For example: `local-zone-app-nyc-1a`.
<5> `machine.openshift.io/zone-group`: Defines the group to where a zone belongs. For example, `us-east-1-nyc-1`.

> <6> `nodeSelector`: Targets edge worker nodes that match the specified labels.

Author: Updated, but I cannot determine what example would suit the remaining `machine.openshift.io/zone-group` array. I see in the following docs a user-replaced value if <infrastructure_ID) was set for this array. Should I set a similar value, because `{ZONE_GROUP_NAME}` is stated twice in the example.

https://docs.openshift.com/container-platform/4.8/machine_management/creating-infrastructure-machinesets.html#machineset-creating_creating-infrastructure-machinesets

The `openshift-installer` creates the machine set manifests file with `node-role.kubernetes.io/edge` and `node-role.kubernetes.io/worker` labels applied to each edge worker node that is located in a Local Zone subnet.

.Prerequisites

* You have an access to the OpenShift CLI (`oc`).
* You deployed your cluster in a Virtual Private Cloud (VPC) with defined Local Zone subnets.
* You ensured that the machine set for the edge workers on Local Zone subnets specifies the taints for `node-role.kubernetes.io/edge`.

.Procedure

. Create a `deployment` resource YAML file for an example application to be deployed in the edge worker node that operates in a Local Zone subnet. Ensure that you specify the correct tolerations that match the taints for the edge worker node.
+
.Example of a configured `deployment` resource for an edge worker node that operates in a Local Zone subnet
[source,yaml]

> ----

Reviewer: Following @yunjiang29's [suggestion](https://github.com/openshift/openshift-docs/pull/57063/files#r1161551637) to add the Namespace to the bundle:

```suggestion
kind: Namespace
apiVersion: v1
metadata:
  name: <local_zone_application_namespace>
---
```

Author: Changed `kind: PersistentVolumeClaim` to `kind: namespace`

Reviewer: @dfitzmau please revert this change, my suggestion was to add the `Namespace` fragment, not to replace. Please see the [discussion](https://github.com/openshift/openshift-docs/pull/57063/files#r1161551637) and [tested commands](https://github.com/openshift/openshift-docs/pull/57063#pullrequestreview-1373220967)

Author: Thanks, @mtulio . I reverted the change and updated the spec according to the [tested commands](https://github.com/openshift/openshift-docs/pull/57063#pullrequestreview-1373220967) link.

.Procedure

. Create a `deployment` resource YAML file for an example application to be deployed in the edge worker node that operates in a Local Zone subnet. Ensure that you specify the correct tolerations that match the taints for the edge worker node.
+
.Example of a configured `deployment` resource for an edge worker node that operates in a Local Zone subnet
[source,yaml]
----
kind: Namespace
apiVersion: v1
metadata:
  name: <local_zone_application_name>
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: <pvc_name>
  namespace: <local_zone_application_namespace>
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: gp2-csi <1>
  volumeMode: Filesystem
---
apiVersion: apps/v1
kind: Deployment <2>
metadata:
  name: <local_zone_application> <3>
  namespace: <local_zone_application_namespace> <4>
spec:
  selector: 
    matchLabels:
      app: <local_zone_application>
  replicas: 1
  template:
    metadata:
      labels: 
        app: <local_zone_application>
        zone-group: ${ZONE_GROUP_NAME} <5>

>     spec:

Reviewer: @dfitzmau you need to add those lines defining the security context to prevent the warning mentioned [here](https://github.com/openshift/openshift-docs/pull/57063/files#r1161553904)

```suggestion
    spec:
      securityContext:
        seccompProfile:
          type: RuntimeDefault
```

@yunjiang29 would you mind validating?

```
cat <<EOF | oc create -f -
kind: Namespace
apiVersion: v1
metadata:
  name: lz-app-ns
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: lz-pvc
  namespace: lz-app-ns
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: gp2-csi 
  volumeMode: Filesystem
---
apiVersion: apps/v1
kind: Deployment 
metadata:
  name: lz-app
  namespace: lz-app-ns 
spec:
  selector:
    matchLabels:
      app: lz-app
  replicas: 1
  template:
    metadata:
      labels:
        app: lz-app
        machine.openshift.io/zone-group: ${ZONE_GROUP_NAME} 
    spec:
      securityContext:
        seccompProfile:
          type: RuntimeDefault
      nodeSelector: 
        machine.openshift.io/zone-group: ${ZONE_GROUP_NAME}
      tolerations: 
      - key: "node-role.kubernetes.io/edge"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
      containers:
        - image: openshift/origin-node
          command:
           - "/bin/socat"
          args:
            - TCP4-LISTEN:8080,reuseaddr,fork
            - EXEC:'/bin/bash -c \"printf \\\"HTTP/1.0 200 OK\r\n\r\n\\\"; sed -e \\\"/^\r/q\\\"\"'
          imagePullPolicy: Always
          name: echoserver
          ports:
            - containerPort: 8080
          volumeMounts:
            - mountPath: "/mnt/storage"
              name: data
      volumes:
      - name: data
        persistentVolumeClaim:
          claimName: lz-pvc
EOF
```

results

```
namespace/lz-app-ns created
persistentvolumeclaim/lz-pvc created
deployment.apps/lz-app created

```
I

Author: Thanks, @mtulio .

I updated the YAML file based on your suggestive addition.

// Module included in the following assemblies:
//
// * post_installation_configuration/cluster-tasks.adoc

ifeval::["{context}" == "installing-aws-localzone"]
:localzone:
endif::[]

:_content-type: PROCEDURE
[id="installation-extend-edge-nodes-aws-local-zones_{context}"]
= Creating user workloads in AWS Local Zones

> After you create an Amazon Web Service (AWS) Local Zone environment, and you deploy your cluster, you can use edge worker nodes to create user workloads in Local Zone subnets. 

Reviewer 6: I believe instances of `worker node` should be changed to `compute node`, unless its in a code block or a parameter. There are a few throughout this PR so I wanted to mention it

Author: Edge worker nodes would be correct, based on already published content in the 4.12 release.

Example of 4.12 content migrated to 4.13: https://docs.openshift.com/container-platform/4.13/installing/installing_aws/installing-aws-localzone.html

After the `openshift-installer` creates the cluster, the installation program automatically specifies a taint effect of `NoSchedule` to each edge worker node. This means that a scheduler does not add a new pod, or deployment, to a node if the pod does not match the specified tolerations for a taint. You can modify the taint for better control over how each node creates a workload in each Local Zone subnet.

The `openshift-installer` creates the machine set manifests file with `node-role.kubernetes.io/edge` and `node-role.kubernetes.io/worker` labels applied to each edge worker node that is located in a Local Zone subnet.

.Prerequisites

> * You have an access to the OpenShift CLI (`oc`).

Reviewer 3: ```suggestion
* You have access to the OpenShift CLI (`oc`).
```

* You deployed your cluster in a Virtual Private Cloud (VPC) with defined Local Zone subnets.

> * You ensured that the machine set for the edge workers on Local Zone subnets specifies the taints for `node-role.kubernetes.io/edge`.

Reviewer 6: I know theres a best practice when writing for machine sets to mention the specific version of this machine set. In this case i believe it should be `compute machine set`
More info can be found in [this doc](https://docs.google.com/document/d/18gQS1sJiuL7LYEqddR8f3CTSNeG-APhrr6srNO5f_jQ/edit)

Author: Thanks. I did not know this doc existed. 

Updated two instances of `machine set` to `compute machine set`.

Author: I wonder if these updates should make their way into the OCP Documentation guidelines?

.Procedure

. Create a `deployment` resource YAML file for an example application to be deployed in the edge worker node that operates in a Local Zone subnet. Ensure that you specify the correct tolerations that match the taints for the edge worker node.
+
.Example of a configured `deployment` resource for an edge worker node that operates in a Local Zone subnet
[source,yaml]
----
kind: Namespace
apiVersion: v1
metadata:

>   name: <local_zone_application_name>

Reviewer 4: @dfitzmau 
```suggestion
  name: <local_zone_application_namespace>
```

// Module included in the following assemblies:
//
// * installing/installing_aws/installing-aws-localzone.adoc

ifeval::["{context}" == "installing-aws-localzone"]
:localzone:
endif::[]

:_content-type: PROCEDURE
[id="installation-extend-edge-nodes-aws-local-zones_{context}"]

> = Creating workloads in AWS Local Zones

Author: Updated to:

> Creating a workload deployment in AWS Local Zones

@mtulio : "special nodes" means modified "edge worker nodes"?

The `openshift-installer` creates the Machine Set manifest and attaches `node-role.kubernetes.io/edge` and `node-role.kubernetes.io/worker` labels to each edge worker node that is located in a Local Zone subnet.

The {product-name} supports using the AWS Load Balancer (ALB) Operator in an existing VPC with configured Local Zone subnets. This means that you can use the Operator to manage an `aws-load-balancer-controller` resource that discovers any Local Zone subnets defined in your existing VPC. You can then create an `Ingress` resource, so that you can expose a running cluster from one Local Zone subnet to other subnets that operate in the same VPC.

.Prerequisites

* You have an access to the OpenShift CLI (`oc`).
* You configured an AWS Local Zones environment. 
* You deployed your cluster om a compatible cloud platform. 
* You installed the AWS Load Balancer (ALB) Operator.
* You created an instance of the AWS Load Balancer Controller.

.Procedure

. Ensure that your worker manifest file for the Local Zone is configured to run workloads. The `zone_group` field specifies your AWS Local Zone that you selected on your AWS account.
+

> .Example of a configured portion of the worker manifest file that the VPC uses to run workloads

Author: SME query: Run, execute or create a workload?

[source,yaml]
----
...
spec:
  metadata:
    labels:
      zone_type: local-zone
      zone_group: ${LZ_ZONE_NAME::-1}
      node-role.kubernetes.io/edge: ""

>   taints:

Reviewer: those labels were changed to: https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R437-R440

Author: Thanks, updated.

      - key: node-role.kubernetes.io/edge
        effect: NoSchedule
...

> spec:

Author: I added two specs. I took the top spec metadata from step G in [Creating a manifest file](https://57063--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-localzone.html#installation-localzone-generate-k8s-manifest_installing-aws-localzone).

For the purpose of this procedure, what `spec:` config is correct?

  nodeSelector:
    zone_group: ${ZONE_GROUP_NAME}
  tolerations:
    - key: "node-role.kubernetes.io/edge"
      operator: "Equal"
      value: ""
      effect: "NoSchedule"
...
----
+
[IMPORTANT]
====
If your worker manifest file does not specify the `node-role.kubernetes.io/edge` toleration, the scheduler cannot schedule pods that have matching taints.
====

. Create an `Ingress` resource YAML file and define your Local Zone subnet metadata in the file. 
+

> .Example of an `Ingress` resource with a defined Local Zone subnet named `ingress-lz-nyc-1`

Author: SME query: Does a user need to define a gateway in this resource? They can use the gateway to toggle private or public access to a subnet?

[source,yaml]
----
...
metadata:
  name: ingress-lz-nyc-1
  namespace: lz-apps
  annotations:
    alb.ingress.kubernetes.io/scheme: internet-facing <1>
    alb.ingress.kubernetes.io/target-type: instance <2>
    alb.ingress.kubernetes.io/subnets: ${SUBNET_ID} <3>
  labels:

>     zone_group: ${ZONE_GROUP_NAME}

Reviewer: the label `zone_group` have been changed

Author: Updated

<3> Specifies the Local Zone subnets that will receive traffic from the ALB Operator. 

. Generate the `Ingress` resource YAML file by issuing the following command in your OpenShift CLI (`oc`):
+
[source,yaml]
----
$ oc create -f <ingress_resource_file_name>
----
+
The ALB Operator reads the `Ingress` resource YAML file and then routes traffic from a Local Zone subnet to other Local Zone subnets that exist in the same VPC. 
+
[IMPORTANT]
====
If you list a high number of Local Zone subnets in your `Ingress` resource file, you might experience latency issues when traffic routes to all the subnets. This behavior occurs if edge worker nodes in a subnet handle data differently than the edge worker node that sent the data. 
====

.Verification

. Verify the status of the `Ingress` resource to show the host of the provisioned AWS Load Balancer (ALB) by running the following command:
+
[source,terminal]
----
HOST=$(kubectl get ingress -n echoserver echoserver -o json | jq -r '.status.loadBalancer.ingress[0].hostname')
----

. Verify the status of the provisioned AWS Load Balancer (ALB) host by running the following command:
+
[source,terminal]
----
$ curl $HOST
----
+

> .Example terminal output that confirms the ALB Operator can route traffic to all subnets defined in the `Ingress` resource

Author: SME query: Is this description correct?

// Module included in the following assemblies:
//
// * installing/installing_aws/installing-aws-localzone.adoc

ifeval::["{context}" == "installing-aws-localzone"]
:localzone:
endif::[]

:_content-type: PROCEDURE
[id="installation-extend-edge-nodes-aws-local-zones_{context}"]
= Extending edge worker nodes into configured AWS Local Zone locations

When installing an {product-title} cluster on Amazon Web Service (AWS), you can use AWS Local Zone subnets to extend regular worker nodes into edge networks. This process creates dedicated edge worker nodes that target specific workloads.

> On completion of the installed cluster, the `openshift-installer` automatically specifies a taint with a `NoSchedule value to each edge worker node. This means that a scheduler does not add a new pod to a node if the pod does not match the taint. Existing pods on the node remain. You can modify the taint value for better control over what workloads can run on each Local Zone subnet.

Reviewer: ```suggestion
On completion of the installed cluster, the `openshift-installer` automatically specifies a taint with a `NoSchedule` value to each edge worker node. This means that a scheduler does not add a new pod to a node if the pod's deployment does not match the tolerations for that taint. Existing pods on the node remain. You can modify the taint value for better control over what workloads can run on each Local Zone subnet.
```

Note sure if we need this:

> Existing pods on the node remain

Author: :heavy_check_mark:

> The `openshift-installer` applies `node-role.kubernetes.io/edge` and `node-role.kubernetes.io/worker` labels to each edge worker node that is located in a Local Zone subnet.

Author: Thanks, @mtulio . How does the following read to you:

> The `openshift-installer` creates the Machine Set manifest and attaches `node-role.kubernetes.io/edge` and `node-role.kubernetes.io/worker` labels to each edge worker node that is located in a Local Zone subnet.

// Module included in the following assemblies:
//
// * installing/installing_aws/installing-aws-localzone.adoc

ifeval::["{context}" == "installing-aws-localzone"]
:localzone:
endif::[]

:_content-type: PROCEDURE
[id="installation-extend-edge-nodes-aws-local-zones_{context}"]
= Creating user workloads in AWS Local Zones

> After you create an Amazon Web Service (AWS) Local Zone environment, and you deploy your cluster, you can use edge worker nodes to create a user workloads in Local Zone subnets. 

Author: @mtulio I've documented the creation of a `deployment` resource and `service` resource. Where do these get created? Inside a cluster that operates in a specific Local Zone subnet? Do the resources apply to all edhe worker nodes that operate inside the cluster?

Reviewer: > Where do these get created? Inside a cluster that operates in a specific Local Zone subnet?
The pods will be created on the `namespace` defined on the deployment and the node (Instance running in Local Zone location/subnet) selected by `nodeSelector`

> Do the resources apply to all edhe worker nodes that operate inside the cluster?

No, only for the nodes matching the `nodeSelector` with tainted nodes matching the `tolerations`.

Author: Thanks. I added the following callout:

> `nodeSelector`: Targets edge worker nodes that match the specified toleration.

> After the `openshift-installer` creates the cluster, the installer automatically specifies a taint effect of `NoSchedule` to each edge worker node. This means that a scheduler does not add a new pod to a node if the pod's deployment does not match the specified tolerations for a taint. You can modify the taint for better control over how each node creates a workload in each Local Zone subnet.

Author: @mtulio . `a pod's deployment` correct? We create a `deployment` resource (a pod) that operates inside an edge worker node?

Reviewer: The `deployment` is an abstract object that defines the pod, the `pod` will run on the 'edge worker nodes' with statement commented above (nodeSelectors and tolerations)

Author: Thanks. I updated this sentence to clarify the pod/deployment query:

> This means that a scheduler does not add a new pod, or deployment, to a node if the pod does not match the specified tolerations for a taint.

The `openshift-installer` creates the Machine Set manifest and attaches `node-role.kubernetes.io/edge` and `node-role.kubernetes.io/worker` labels to each edge worker node that is located in a Local Zone subnet.

.Prerequisites

> * You have an access to the OpenShift CLI (`oc`).

Author: @mtulio . I should mention the creation of the worker manifest file here and the importance of defining the taint?

Reviewer: > * You configured an AWS Local Zones environment. 

I think we must improve this line with something like a cluster deployed extending works to Local Zones (currently day-0 docs).

With regards the `nd the importance of defining the taint`, not sure about defining the importance as it should be a concept of K8s. The taints is a design choice, the user can deploy a cluster without taints and tolerations, but we decide that architecture to avoid spreading cluster components in more expensive locations, impacting/increasing in costs and latency.

Also if the user followed our documentation, the worker nodes must be tainted, so maybe is worth to mentioned that we expected the nodes to be tainted to match the tolerations mentioned on the use case/deployment.
wdyt?

Author: Thanks, @mtulio . Yes, good suggestion. I also removed a bullet point, because it melds with my new bullet point.

I updated the bullet point to state:

>  You deployed your cluster in a Virtual Private Cloud (VPC) with defined Local Zone subnets for extending regular worker nodes to edge worker nodes

* You ensured that the machine set for the edge workers on Local Zone subnets specifies the taints for `node-role.kubernetes.io/edge`.

.Procedure

. Create a `deployment` resource YAML file for an example application to be deployed in the edge worker node that operates in a Local Zone subnet. Ensure that you specify the correct tolerations that match the taints for the edge worker node.
+
.Example of a configured `deployment` resource for an edge worker node that operates in a Local Zone subnet
[source,yaml]
----
apiVersion: apps/v1
kind: Deployment <1>
metadata:

>   name: <local_zone_name> <2>
>   namespace: <local_workload_zoneName> <3>

Reviewer: ```suggestion
  name: <local_zone_application> <2>
  namespace: <local_zone_application_namespace> <3>
```

  namespace: <local_zone_namespace> <3>
spec:
  selector: 
    matchLabels:
      app: <local_zone_application>
  replicas: 1
  template:
    metadata:
      labels: 
        app: <local_zone_application>
        machine.openshift.io/zone-group: ${ZONE_GROUP_NAME} <4>
    spec:
      nodeSelector:
        machine.openshift.io/zone-group: ${ZONE_GROUP_NAME}

>       tolerations: <5>

Author: @mtulio . Where does one define a taint? For example: https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R437-R440

Reviewer: the node is created with the taint defined on the MachineSet (linked by you).
This toleration must match with the taints to allow scheduling the pod on the edge worker node.

Author: Updated to 

>  Sets the tag where the scheduler can apply a taint to affected nodes. Toleration must match the taint that you defined in the machine set manifest file.

      - key: "node-role.kubernetes.io/edge"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
      containers:
        - image: openshift/origin-node
          command:
           - "/bin/socat"
          args:
            - TCP4-LISTEN:8080,reuseaddr,fork
            - EXEC:'/bin/bash -c \"printf \\\"HTTP/1.0 200 OK\r\n\r\n\\\"; sed -e \\\"/^\r/q\\\"\"'
          imagePullPolicy: Always
          name: echoserver
          ports:
            - containerPort: 8080
----
<1> `kind`: Defines the `deployment` resource.
<2> `name`: Specifies the name of the service. 
<3> `namespace:` Defines the namespace that the `service` resource targets.
<4> `machine.openshift.io/zone-group`: Defines the group to where a zone belongs. For example, `us-east-1-nyc-1` indicates the region code, `nyc`, and identifiers for a Local Zone.
<5> `tolerations`: Sets the tag where the scheduler can apply a taint to affected nodes.

> . Create a `service` resource YAML file for the node.

Author: @mtulio . What is the overall purpose of the `service` resource. After its creation does some mechanism, such as an Operator, use it?

Reviewer: Service will expose the pod to the network. So basically is the way to expose one App/API to the other services running on the cluster, and eventually external when properly set (for example using ALB as ingress).

[1] https://kubernetes.io/docs/concepts/services-networking/service
[2] https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types

Author: Thanks!

I updated the content to state:

> This resource exposes a pod from a targeted edge worker node to your Local Zone network.

.Example of a configured `service` resource for an edge worker node that operates in a Local Zone subnet
[source,yaml]
----
apiVersion: v1
kind: Service <1>
metadata:
  name:  <local_zone_service_name>
  namespace: <local_zone_name>
spec:
  ports:
    - port: 80
      targetPort: 8080
      protocol: TCP
  type: NodePort
  selector: <2>
    app: local-zone-demo-app-nyc-1
----
<1> `kind`: Defines the `service` resource.

> <2> `selector:` Specifies the label type applied to managed pods.

Author: @mtulio .

After a user creates this resource, what happens next? What mechanism does the rest o the tasks? Must a user do more, or does something automatically complete the next steps?

Reviewer: As commented on Slack, we can stop here and mention, or put a short paragraph of possibilities (but not required) that users could **look** next, to avoid dup content, of ALB Operator.

Author: Thanks for pointing this out. I added the following sentence to callout <5>:

> The `key:` value must match the taint value that you specified in the worker manifest file.

Author: Updated.

>   name:  <local_zone_service_name>
>   namespace: <local_zone_name>

Reviewer: same comment above, try to follow the same standard of the Deployment for `namespace`. For the name, can be anything considering the `selector` is correct, but for didactic reasons we could (open to hear from your based in other examples from our docs) create the same name of the app, so copying the Deployment suggestion:

```suggestion
  name: <local_zone_applicatio>
  namespace: <local_zone_application_namespace>
```

Reviewer: @dfitzmau you missed the namespace that should match with https://github.com/openshift/openshift-docs/pull/57063/files#diff-6684ff2ad223a0c07c100d71fe49840776acf869d0d2873b6df6675abd3a6684R35

Author: Thanks for pointing this out. Not sure why this change was not added to the commit.

// Module included in the following assemblies:
//
// * post_installation_configuration/cluster-tasks.adoc

ifeval::["{context}" == "installing-aws-localzone"]
:localzone:
endif::[]

:_content-type: PROCEDURE
[id="installation-extend-edge-nodes-aws-local-zones_{context}"]
= Creating user workloads in AWS Local Zones
After you create an Amazon Web Service (AWS) Local Zone environment, and you deploy your cluster, you can use edge worker nodes to create a user workloads in Local Zone subnets. 

After the `openshift-installer` creates the cluster, the installer automatically specifies a taint effect of `NoSchedule` to each edge worker node. This means that a scheduler does not add a new pod, or deployment, to a node if the pod does not match the specified tolerations for a taint. You can modify the taint for better control over how each node creates a workload in each Local Zone subnet.

> The `openshift-installer` creates the machine set manifest file and attaches `node-role.kubernetes.io/edge` and `node-role.kubernetes.io/worker` labels to each edge worker node that is located in a Local Zone subnet.

Reviewer: ```suggestion
The `openshift-installer` creates the machine set manifest manifests with `node-role.kubernetes.io/edge` and `node-role.kubernetes.io/worker` labels to each edge worker node that is located in a Local Zone subnet.
```

Author: Updated to:

> The `openshift-installer` creates the machine set manifests file with `node-role.kubernetes.io/edge` and `node-role.kubernetes.io/worker` labels applied to each edge worker node that is located in a Local Zone subnet.

I added the "applied" verb, because some action is being done to edge worker nodes.

.Prerequisites

* You have an access to the OpenShift CLI (`oc`).
* You deployed your cluster in a Virtual Private Cloud (VPC) with defined Local Zone subnets for extending regular worker nodes to edge worker nodes. 
* You ensured that your machine set manifest file specifies `node-role.kubernetes.io/edge` for the `taint` parameter.

.Procedure

. Create a `deployment` resource YAML file for an edge worker node that operates in a Local Zone subnet. Ensure that you specify the correct tolerations for your edge worker node, so that the scheduler can schedule any pods with matching taints to the cluster.
+
.Example of a configured `deployment` resource for an edge worker node that operates in a Local Zone subnet
[source,yaml]
----
apiVersion: apps/v1
kind: Deployment <1>
metadata:
  name: <local_zone_name> <2>

>   namespace: <local_zone_namespace> <3>

Reviewer: this could cause confusion, IMHO. according to the original example, it can be anything user-defined (free form). I would keep the "sample values" add update the legend that those values can be changed. Keeping "sample-..."  or something related (maybe "my-app...") could be more helpful (or maybe update from `<local_zone_name>` to `<user_workload_zoneName>`.
Reading that statement "<>" I felt like something mandatory related to Local zone deployment, and it's not. The intent is to provide an example to use. WDYT?

Author: Yeah, we try and use user-replaced values to avoid any scenarios where a user might assume values in an example is what they need to define in their config file. Users should be aware of their use in our docs as they are common ways to call out that the values are examples and need to be replaced. 

See https://redhat-documentation.github.io/supplementary-style-guide/#user-replaced-values

spec:
  selector: 
    matchLabels:
      app: <local_zone_application>
  replicas: 1
  template:
    metadata:
      labels: 
        app: <local_zone_application>
        machine.openshift.io/zone-group: ${ZONE_GROUP_NAME} <4>
    spec:
      nodeSelector: <5>
        machine.openshift.io/zone-group: ${ZONE_GROUP_NAME}
      tolerations: <6>
      - key: "node-role.kubernetes.io/edge"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
      containers:
        - image: openshift/origin-node
          command:
           - "/bin/socat"
          args:
            - TCP4-LISTEN:8080,reuseaddr,fork
            - EXEC:'/bin/bash -c \"printf \\\"HTTP/1.0 200 OK\r\n\r\n\\\"; sed -e \\\"/^\r/q\\\"\"'
          imagePullPolicy: Always
          name: echoserver
          ports:
            - containerPort: 8080
----
<1> `kind`: Defines the `deployment` resource.
<2> `name`: Specifies the name of the service. 

> <3> `namespace:` Defines the namespace that the `service` resource targets.

Author: Sorry, @mtulio . What callout does your comment refer to?

> <3> `namespace:` Defines the namespace that the `service` resource targets.

Reviewer: Sorry, that line is confusing imho. The `service` is one object on the `namespace`. 

I commented above https://github.com/openshift/openshift-docs/pull/57063/files#r1147961727

The idea of this book is to explain an user-workload to run on Local Zone node/feature. So the app name and it's namespace can be anything the user wants, we can provide user-wide values, and some examples. "foo-app-zone-use1-nyc-1a" [...]

> <3> `namespace:` Defines the namespace that the `service` resource targets.

Reviewer: wow, you are right! thanks for pointing it out! thanks!

> <4> `machine.openshift.io/zone-group`: Defines the group to where a zone belongs. For example, `us-east-1-nyc-1` indicates the region code, `nyc`, and identifiers for a Local Zone.

Reviewer: ```suggestion
<4> `machine.openshift.io/zone-group`: Defines the group to where a zone belongs. For example, `us-east-1-nyc-1`.
```

<5> `nodeSelector`: Targets edge worker nodes that match the specified toleration.

> <6> `tolerations`: Sets the tag where the scheduler can apply a taint to affected nodes. The `key:` value must match the taint value that you specified in the machine set manifest file.

Reviewer: IMHO you don't need to explain what the toleration/taint is, or you can specify our documentation for it[1]. See how logging is used it[2]:
```
- This toleration is added to the Elasticsearch pods.
```

[1] https://docs.openshift.com/container-platform/4.12/nodes/scheduling/nodes-scheduler-taints-tolerations.html
[2] https://docs.openshift.com/container-platform/4.12/logging/config/cluster-logging-tolerations.html

Author: I could add a link, but the short callout would help for the purposes of this section and the example we're trying to detail.

Reviewer: @dfitzmau my proposal was not to add the link, was to show how our documentation is used when referring to tolerations, a short sentence would be preferable as our goal is not to explain what toleration is, but that we are using it.
I suggest being direct too, something like:
```
<6> `tolerations`: Sets the values that match with the `taints` defined on the MachineSet manifest for the Local Zone node.
```

Author: Updated.

Should I remove the following links?

 * xref:../nodes/scheduling/nodes-scheduler-taints-tolerations.adoc[Understanding taints and tolerations]

* xref:../logging/config/cluster-logging-tolerations.adoc[Using tolerations to control OpenShift Logging pod placement]
